ivadomed/loader/dataset_group.py

```diff
@@ -0,0 +1,295 @@
+        self.n_expected_input: int = int(dict_file_group_spec.get(DataloaderKW.EXPECTED_INPUT, 0))
+        if self.n_expected_input <= 0:
+            raise ValueError("At DataSetGroup level, Number of Expected Input must be > 0")
+
+        self.n_expected_gt: int = int(dict_file_group_spec.get(DataloaderKW.EXPECTED_GT, 0))
+        if self.n_expected_gt <= 0:
+            raise ValueError("At DataSetGroup level, Number of Expected Ground Truth must be > 0")
+
+        # This is the key attribute that needs to be populated once data loading is complete.
+        if (ModelParamsKW.IS_2D not in config.model_params and not config.model_params.get(ModelParamsKW.IS_2D)):
+            self.validate_update_3d_train_val_test_and_filename_pairs(config, dict_file_group_spec)
+        else:
+            self.validate_update_2d_train_val_test_dataset_and_filename_pairs(dict_file_group_spec)
+
+        # Assign default labels if not provided.
+        self.name: str = dict_file_group_spec.get(DataloaderKW.DATASET_GROUP_LABEL, "DefaultDatasetGroupLabel")
+
+    def validate_update_3d_train_val_test_and_filename_pairs(self, config, dict_file_group_spec):
+        # Instantiate all the 3D Train/Val/Test Datasets
+        for a_train_dataset in dict_file_group_spec[DataloaderKW.TRAINING]:
+            if a_train_dataset[DataloaderKW.TYPE] == "FILES":
+                train_set = Files3DDataset(a_train_dataset, config)
+                self.train_filename_pairs.extend(train_set.filename_pairs)
+        for a_val_dataset in dict_file_group_spec[DataloaderKW.VALIDATION]:
+            if a_val_dataset[DataloaderKW.TYPE] == "FILES":
+                val_set = Files3DDataset(a_val_dataset, config)
+                self.val_filename_pairs.extend(val_set.filename_pairs)
+        for a_test_dataset in dict_file_group_spec[DataloaderKW.TEST]:
+            if a_test_dataset[DataloaderKW.TYPE] == "FILES":
+                test_set = Files3DDataset(a_test_dataset, config)
+                self.test_filename_pairs.extend(test_set.filename_pairs)
+
+    def validate_update_2d_train_val_test_dataset_and_filename_pairs(self, dict_file_group_spec):
+        # Instantiate all the 2D Train/Val/Test Datasets
+
+        # Given a JSON file, try to load the file pairing from it.
+        assert dict_file_group_spec.get(DataloaderKW.TYPE).upper() == "FILES", \
+            f"Invalid DataLoader type specified: {dict_file_group_spec.get(DataloaderKW.TYPE)}"
+
+        # handle simple json specifications
+        self.train_filename_pairs = self.parse_simple_spec_dict_for_filename_pairs(
+            dict_file_group_spec.get(DataloaderKW.TRAINING)
+        )
+
+        self.val_filename_pairs = self.parse_simple_spec_dict_for_filename_pairs(
+            dict_file_group_spec.get(DataloaderKW.VALIDATION)
+        )
+
+        self.test_filename_pairs = self.parse_simple_spec_dict_for_filename_pairs(
+            dict_file_group_spec.get(DataloaderKW.TEST)
+        )
```

Comment:
From my tests, it seems that with the current implementation, all 3 datasets (train/valid/test) must be present for each dataset_group or it throws an error:
```
Traceback (most recent call last):
  File "/home/mhbourget/venv-ivadomed-296/bin/ivadomed", line 11, in <module>
    load_entry_point('ivadomed', 'console_scripts', 'ivadomed')()
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 872, in run_main
    run_command(context=context,
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 535, in run_command
    execute_multi_path_training(
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 680, in execute_multi_path_training
    all_data = AllDatasetGroups(
  File "/home/mhbourget/code/ivadomed/ivadomed/loader/all_dataset_group.py", line 56, in __init__
    dataset_group = FileDatasetGroup(dict_dataset_group_spec, config)
  File "/home/mhbourget/code/ivadomed/ivadomed/loader/dataset_group.py", line 61, in __init__
    self.validate_update_2d_train_val_test_dataset_and_filename_pairs(dict_file_group_spec)
  File "/home/mhbourget/code/ivadomed/ivadomed/loader/dataset_group.py", line 93, in validate_update_2d_train_val_test_dataset_and_filename_pairs
    self.val_filename_pairs = self.parse_simple_spec_dict_for_filename_pairs(
  File "/home/mhbourget/code/ivadomed/ivadomed/loader/dataset_group.py", line 160, in parse_simple_spec_dict_for_filename_pairs
    if a_simple_loader_json.get(DataloaderKW.MISSING_FILES_HANDLE) == FileMissingHandleKW.SKIP:
AttributeError: 'NoneType' object has no attribute 'get'
```

From previous discussion (https://github.com/ivadomed/ivadomed/pull/1201#pullrequestreview-1186663781), I think we should be able to list only 1 or 2 of these datasets per dataset_group.
ex: train and validation comes from a the dataset group 1, and test comes from the dataset group 2.
Let me know if that makes sense or if I'm mistaken.

Reply:
Excellent point.  I now build a early return for the validation process when the Dict is none and also carry out an internal check for when al three are missing. Thanks for bringing up this excellent point.

Reply:
Thanks! It's working well now. I think we would also need to add a check at a higher level (all dataset or consolitdation?) in case one or more of the datasets (train/valid or test) is missing, which does not seem to be implemented at the moment and end up with an out of range error:
```
Traceback (most recent call last):
  File "/home/mhbourget/venv-ivadomed-296/bin/ivadomed", line 11, in <module>
    load_entry_point('ivadomed', 'console_scripts', 'ivadomed')()
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 870, in run_main
    run_command(context=context,
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 533, in run_command
    execute_multi_path_training(
  File "/home/mhbourget/code/ivadomed/ivadomed/main.py", line 688, in execute_multi_path_training
    ds_valid = ConsolidatedDataset.consolidate_AllDatasetGroups_to_a_specific_filedataset_type(all_data, DataloaderKW.VALIDATION)
  File "/home/mhbourget/code/ivadomed/ivadomed/loader/consolidation.py", line 93, in consolidate_AllDatasetGroups_to_a_specific_filedataset_type
    return ConsolidatedDataset(filename_pairs, all_dataset_groups.config)
  File "/home/mhbourget/code/ivadomed/ivadomed/loader/consolidation.py", line 45, in __init__
    super().__init__(
  File "/home/mhbourget/code/ivadomed/ivadomed/loader/mri2d_segmentation_dataset.py", line 118, in __init__
    self.n_contrasts = len(self.filename_pairs[0][0])
IndexError: list index out of range
```